### PR TITLE
streaming multipack for pretraining dataset

### PR DESCRIPTION
**Pull request by jinwonkim93**
This pull request introduces support for multipacking in streaming pretraining datasets.
Due to the immense size of these datasets, traditional methods of loading them entirely into memory are not feasible. 
The proposed solution aims to enhance efficiency and scalability.

need guide for making config.
this multipack does not use BatchSamplerDataCollatorForSeq2Seq, only use DataCollatorForSeq2Seq due to huggingface dataset map function.

### CODE REVIEW

File: src/axolotl/utils/data.py [hunk 1]
```diff
@@ -819,3 +824,68 @@ def load_pretraining_dataset(path, tokenizer, max_tokens=2048, seed=42):
         remove_columns=dataset.features.keys(),
     )
     return dataset
+
+
+def encode_packed_pretraining(
+    tokenizer: PreTrainedTokenizerBase,
+    examples: List[str],
+    max_seq_length: int = 8192,
+    sample_packing_efficiency: int = 1,
```
Comment (winglian) [inline src/axolotl/utils/data.py]:
```suggestion
    sample_packing_efficiency: float = 1.0,
```

**Comment by winglian** [comment]
Here's a patch file I used to test a c4 pretraiing dataset with tinnyllama. multigpu doesn't work currently w this since I think it needs a proper data collator to pad the samples to the same sequence length
[patch0.patch](https://github.com/OpenAccess-AI-Collective/axolotl/files/13837588/patch0.patch)

**Comment by casper-hansen** [comment]
Would this streaming feature work with S3, GCS, Azure Blob Storage?

**Comment by winglian** [comment]
This PR is ready for review and should resolve #1026. @mhenrichsen

**Comment by mhenrichsen** [comment]
Confirmed working on single gpu. Currently fails on multi gpu.

**Pull request merged**